### PR TITLE
Upgrade to pydantic v2

### PR DESCRIPTION
**Pull request by huard**
Updates the code to use pydantic v2.  This is large PR with lots of changes and a fairly high probability of breakage.

### CODE REVIEW

**Comment by huard** [comment]
Need to wait on https://github.com/hydroshare/hsmodels/pull/45 to merge. A notebook depends on `hsclient` and the upgrade to v2 breaks the doc build.

**Comment by huard** [comment]
I suggest to disable the notebook using hsclient and merge this. We should be able to include it back once hsclient releases a version with support for pydantic v2.

File: docs/notebooks/index.rst [hunk 1]
```diff
@@ -37,7 +37,6 @@ Advanced workflows
    Hydrological_realtime_forecasting
    Assess_probabilistic_flood_risk
    Distributed_hydrological_modelling
-   HydroShare_integration
```
[the remark this answers is not in the document]
Comment (huard) [reply docs/notebooks/index.rst:40]: Not obvious how to comment stuff out of the toctree

**Review by Zeitsperre — APPROVED** [review]
The build and documentation is now passing. Is there anything in particular that needs to be looked over? Otherwise, I trust that your pydantic v2 implementation is fine.

**Comment by huard** [comment]
Thanks ! The changes are so large, than I'm counting on the test suite to catch problems, really. There would probably be benefits to improve the coverage of the test suite before the next release. 

We don't have code coverage up and running for this repo, right ? 

I think my suggestion is to merge this, look at the code coverage and try to improve it for critical areas before the next release.

**Comment by Zeitsperre** [comment]
@huard

 We _do_ have code coverage for this repo (https://coveralls.io/github/CSHS-CWRA/RavenPy : Coverage currently at 82%) but it doesn't seem to be reporting properly on Pull Requests. Weird.

I've resynced the webhook, so maybe that fixed it? Merging now.

**Pull request merged**